### PR TITLE
[test][tensorflow|pytorch][training] Add SM Profiler Integration Tests

### PR DESCRIPTION
**Pull request by arjkesh**
*Issue #, if available:*

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [x] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [x] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [x] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [x] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [x] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [x] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [x] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [x] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
- Add new section for release candidate integration tests
- Implement SMProfiler RC integration tests

*Tests run:*
- Tested SMProfiler RC integration tests on PR

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/config/test_config.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 DISABLE_EFA_TESTS = False
 
 DISABLE_SANITY_TESTS = False
-DISABLE_SAGEMAKER_TESTS = False
-DISABLE_ECS_TESTS = False
-DISABLE_EKS_TESTS = False
-DISABLE_EC2_TESTS = False
+DISABLE_SAGEMAKER_TESTS = True
```
Comment (arjkesh) [inline src/config/test_config.py:11]: temporary changes

File: src/config/build_config.py [hunk 2]
```diff
@@ -5,9 +5,9 @@
 # Do remember to revert it back to False before merging any PR (including NEURON dedicated PR)
 ENABLE_NEURON_MODE = False
 # Frameworks for which you want to disable both builds and tests
-DISABLE_FRAMEWORK_TESTS = []
```
Comment (arjkesh) [inline src/config/build_config.py:8]: temporary changes

File: pytorch/buildspec.yml [hunk 3]
```diff
@@ -67,32 +67,32 @@ images:
...
-  #   python_version: &DOCKER_PYTHON_VERSION py3
-  #   tag_python_version: &TAG_PYTHON_VERSION py36
-  #   cuda_version: &CUDA_VERSION cu111
-  #   os_version: &OS_VERSION ubuntu18.04
-  #   tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION ]
-  #   docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.,
-  #                        *DEVICE_TYPE ]
-  #   context:
-  #     <<: *TRAINING_CONTEXT
+   BuildCPUPTTrainPy3DockerImage:
```
Comment (arjkesh) [inline pytorch/buildspec.yml:70]: temporary changes

File: testspec.yml [hunk 4]
```diff
@@ -9,9 +9,11 @@ phases:
       - start-dockerd
       - pip install -r src/requirements.txt
       - python src/send_status.py --status 2
+      - apt-get update && apt-get install unzip
   build:
     commands:
       - echo Logging in to Amazon ECR...
+      - export TEST_TYPE="release_candidate_integration"
```
Comment (arjkesh) [inline testspec.yml:16]: temporary changes

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    # Command to set all necessary environment variables
+    export_cmd = " && ".join(f"export {key}={val}" for key, val in spec_file_envs.items())
+    export_cmd = f"{export_cmd} && export ENV_{processor.upper()}_TRAIN_IMAGE={image}"
+
+    test_results_outfile = os.path.join(os.getcwd(), f"{get_container_name('smprof', image)}.txt")
+    with ctx.prefix(f"cd {profiler_tests_dir}"):
+        with ctx.prefix(f"cd sagemaker-tests/tests && {export_cmd}"):
+            test_output = ctx.run(
+                f"pip install smdebug && pytest --json-report --json-report-file={test_results_outfile} -n=auto -v -s -W=ignore {test_file}::test_{processor}_jobs",
```
Comment (tejaschumbalkar) [inline test/dlc_tests/release_candidate_integration/test_sm_profiler.py:119]: Are we installing `smdebug` on the host machine? why is that required?

**Review by tejaschumbalkar — COMMENTED** [review]
At what stage the test will be ran in mainline context?

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    # Install tf datasets
+    ctx.run(
+        f"echo 'tensorflow-datasets==4.0.1' >> {profiler_tests_dir}/sagemaker-tests/tests/scripts/tf_scripts/requirements.txt",
+        hide=True,
+    )
+
+    run_sm_profiler_tests(tensorflow_training, profiler_tests_dir, "test_profiler_tensorflow.py", processor)
+
+
+def run_sm_profiler_tests(image, profiler_tests_dir, test_file, processor):
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:81]: yeah most definitely, will add

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    smdebug_version = ctx.run(
+        f"docker run {image} python -c 'import smdebug; print(smdebug.__version__)'", hide=True
+    ).stdout.strip()
+
+    if Version(smdebug_version) < Version("1"):
+        pytest.skip(f"smdebug version {smdebug_version} is less than 1, so smprofiler not expected to be present")
+
+    # Install smprofile requirements from GitHub
+    ctx.run(
+        f"pip install -r https://raw.githubusercontent.com/awslabs/sagemaker-debugger/master/config/profiler/requirements.txt",
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:92]: I had this prior but needed to update because old branches have a broken requirements.txt, so unfortunately we need to use master for now. I can add a comment for this

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py (hunk 5 above)
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:119]: This may no longer be required, but in the original script, smdebug is a test requirement, but not specified in their test requirements explicitly. I can try without it

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    ctx.run(f"mkdir -p {profiler_tests_dir}")
+
+    # Download sagemaker-tests zip
+    sm_tests_zip = "sagemaker-tests.zip"
+    ctx.run(f"aws s3 cp s3://smprofiler-test-artifacts/{sm_tests_zip} {profiler_tests_dir}/{sm_tests_zip}", hide=True)
+    ctx.run(f"cd {profiler_tests_dir} && unzip {sm_tests_zip}", hide=True)
+
+    # Install tf datasets
+    ctx.run(
+        f"echo 'tensorflow-datasets==4.0.1' >> {profiler_tests_dir}/sagemaker-tests/tests/scripts/tf_scripts/requirements.txt",
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:74]: This is unfortunately a test requirement from their test runner - I can keep this as a variable for ease of change

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    # Download sagemaker-tests zip
+    sm_tests_zip = "sagemaker-tests.zip"
+    ctx.run(f"aws s3 cp s3://smprofiler-test-artifacts/{sm_tests_zip} {profiler_tests_dir}/{sm_tests_zip}")
+
+    # PT test setup requirements
+    with ctx.prefix(f"cd {profiler_tests_dir}"):
+        ctx.run(f"unzip {sm_tests_zip}")
+        with ctx.prefix("cd tests/scripts/pytorch_scripts"):
+            ctx.run("mkdir -p data")
+            ctx.run("aws s3 cp s3://smdebug-testing/datasets/cifar-10-python.tar.gz data/cifar-10-batches-py.tar.gz")
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:43]: I believe so, will confirm

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    ctx = Context()
+
+    profiler_tests_dir = os.path.join(
+        os.getenv("CODEBUILD_SRC_DIR"), get_container_name("smprof", pytorch_training), "smprofiler_tests"
+    )
+    ctx.run(f"mkdir -p {profiler_tests_dir}")
+
+    # Download sagemaker-tests zip
+    sm_tests_zip = "sagemaker-tests.zip"
+    ctx.run(f"aws s3 cp s3://smprofiler-test-artifacts/{sm_tests_zip} {profiler_tests_dir}/{sm_tests_zip}")
```
[the remark this answers is not in the document]
Comment (tejaschumbalkar) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:36]: Lets add a TODO to change the bucket once the smprofiler team has the bucket ready

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        yml_envs = yaml.safe_load(sf)
+        spec_file_envs = yml_envs.get("env", {}).get("variables")
+
+    # Command to set all necessary environment variables
+    export_cmd = " && ".join(f"export {key}={val}" for key, val in spec_file_envs.items())
+    export_cmd = f"{export_cmd} && export ENV_{processor.upper()}_TRAIN_IMAGE={image}"
+
+    test_results_outfile = os.path.join(os.getcwd(), f"{get_container_name('smprof', image)}.txt")
+    with ctx.prefix(f"cd {profiler_tests_dir}"):
+        with ctx.prefix(f"cd sagemaker-tests/tests && {export_cmd}"):
```
Comment (tejaschumbalkar) [inline test/dlc_tests/release_candidate_integration/test_sm_profiler.py:117]: What type of test would be executed? SageMaker or EC2?

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py (hunk 5 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:119]: do we need a rerun here?

File: test/dlc_tests/release_candidate_integration/test_sm_profiler.py (hunk 11 above)
Comment (arjkesh) [reply test/dlc_tests/release_candidate_integration/test_sm_profiler.py:117]: will be SM

File: test/testrunner.py [hunk 12]
```diff
@@ -316,7 +316,9 @@ def main():
         sm_utils.generate_empty_report(report, test_type, "huggingface")
         return
 
-    if specific_test_type in ("sanity", "ecs", "ec2", "eks", "canary", "bai", "quick_checks"):
+    if specific_test_type in (
+            "sanity", "ecs", "ec2", "eks", "canary", "bai", "quick_checks", "release_candidate_integration"
```
Comment (SergTogul) [inline test/testrunner.py:320]: could you also add _"release_candidate_integration"_ to if block at line 313?

Comment (arjkesh) [reply test/testrunner.py:321]: Added, thanks for pointing that out

Comment (SergTogul) [reply test/testrunner.py:321]: Sorry, didn't push "Add comment"
I don't see any handlers for "release_candidate_integration" down there. Is it supposed to be added later on?

Comment (arjkesh) [reply test/testrunner.py:321]: Not sure I understand - are you referring to the finally block around line 422? There shouldn't need to be any specific RC integration handlers here afaik

Comment (arjkesh) [reply test/testrunner.py:321]: Oh, maybe you mean why do we not have any "if" conditions for these tests? This is because they run with the default behavior 

```
pytest_cmd = ["-s", "-rA", test_path, f"--junitxml={report}", "-n=auto"]
```

so there is no need to add

Comment (SergTogul) [reply test/testrunner.py:321]: From 334 till 407 we're creating a pytest command based on test type and other parameters. Since you don't have a special "if" block for "release_candidate_integration", you'll end up with a command created at [402](https://github.com/aws/deep-learning-containers/pull/1178/files#diff-7cbc819cfee09f8be65c72606589778a1d44839986c87dfb0930a40ab75820efR402). 
So if you need any special/custom parameter to be added to the command you'll need an additional handler.

File: testspec.yml [hunk 13]
```diff
@@ -9,6 +9,7 @@ phases:
       - start-dockerd
       - pip install -r src/requirements.txt
       - python src/send_status.py --status 2
+      - apt-get update && apt-get install unzip
```
Comment (tejaschumbalkar) [inline testspec.yml:12]: Lets install the `unzip` package on the codebuild image itself and also `apt-get update` if required

File: test/testrunner.py (hunk 12 above)
Comment (arjkesh) [reply test/testrunner.py:321]: Nothing else is needed, default path is fine as mentioned above. Thanks for checking this

**Review by SergTogul — APPROVED** [review]

**Review by tejaschumbalkar — APPROVED** [review]
